src/stream.js

```diff
@@ -0,0 +1,488 @@
+var q = require('./query')
+var util = require('./_util')
+
+var DefaultEvents = ['start', 'error', 'version', 'history_rewrite']
+var DocumentStreamEvents = DefaultEvents.concat(['snapshot'])
+
+/**
+ * The internal stream client interface. This class handles the network side of
+ * a stream subscription. This implementation is environment aware. It attempts
+ * to use the native `fetch` API on browsers, while relying on the underlying
+ * `cross-fetch` polyfill from {@link Client} on NodeJS.
+ *
+ * Known limitations:
+ * * NodeJS uses HTTP1;
+ * * Only modern browsers are supported.
+ *
+ * @constructor
+ * @param {Client} client The FaunaDB client.
+ * @param {module:query~ExprArg} expression The fauna expression to subscribe.
+ * @param {module:stream~Options} options The stream options.
+ * @param {function} onEvent The stream event's callback.
+ * @private
+ */
+function StreamClient(client, expression, options, onEvent) {
+  options = util.applyDefaults(options, {
+    fields: null,
+  })
+
+  this._client = client
+  this._onEvent = onEvent
+  this._query = q.wrap(expression)
+  this._urlParams = options.fields ? { fields: options.fields.join(',') } : null
+  this._fetch = platformCompatibleFetchOverride()
+  this._abort = new AbortController()
+  this._state = 'idle'
+
+  // Determines a suitable override for the `cross-fetch` polyfill in the
+  // current platform. If the `fetch` API is not supported, returns a function
+  // that fails any requests so that errors are dispatched to appropriate error
+  // handlers rather than introducing the need for a try/catch.
+  function platformCompatibleFetchOverride() {
+    var httpClient = client._http
+    var fetch = httpClient._fetch
+
+    if (util.isNodeEnv() || fetch.override || !fetch.polyfill) {
+      return null // no override needed
+    }
+
+    fetch = http.resolveFetch(null, false)
+    if (fetch === null) {
+      fetch = function() {
+        return Promise.reject(
+          new errors.StreamsNotSupported(
+            'Could not find a stream compatible fetch function. ' +
+              'Please, consider providing a Fetch API compatible function ' +
+              'with streamable response bodies.'
+          )
+        )
+      }
+    }
+
+    return fetch
+  }
+}
+
+/**
+ * Takes a snapshot of the current query. Assumes the subscribed query returns a
+ * reference.
+ */
+StreamClient.prototype.snapshot = function() {
+  var self = this
+  self._client
+    .query(q.Get(self._query))
+    .then(function(doc) {
+      self._onEvent({
+        event: 'snapshot',
+        data: doc,
+      })
+    })
+    .catch(function(error) {
+      self._onEvent({
+        event: 'error',
+        data: error,
+      })
+    })
+}
+
+/** Initiates the stream subscription.  */
+StreamClient.prototype.subscribe = function() {
+  var self = this
+  if (self._state === 'idle') {
+    self._state = 'open'
+  } else {
+    throw new Error('The stream alread started at once.')
```

Comment:
`already`

Reply:
"At once" is a synonym for "immediately". Is that what the error message is trying to convey? Or do you mean "The stream has already been started." ?

Reply:
The latter.